itests/kit/tools_eudico.go

```diff
@@ -91,67 +100,117 @@ func WaitForBalance(ctx context.Context, addr addr.Address, balance uint64, api
 	for big.Cmp(currentBalance, targetBalance) != 1 {
 		select {
 		case <-ctx.Done():
-			return xerrors.New("closed channel")
+			return fmt.Errorf("closed channel")
 		case <-ticker.C:
 			currentBalance, err = api.WalletBalance(ctx, addr)
 			if err != nil {
 				return err
 			}
 		case <-timer:
-			return xerrors.New("balance timer exceeded")
+			return fmt.Errorf("balance timer exceeded")
 		}
 	}
 
 	return nil
 }
 
-func SubnetPerformHeightCheckForBlocks(ctx context.Context, validatedBlocksNumber int, subnetAddr addr.SubnetID, api napi.FullNode) error {
-	subnetHeads, err := getSubnetChainHead(ctx, subnetAddr, api)
+// SubnetMinerMinesBlocks checks that the miner can mine some `m` blocks of `n` all blocks in the subnet.
+func SubnetMinerMinesBlocks(ctx context.Context, m, n int, subnetAddr addr.SubnetID, miner addr.Address, api napi.FullNode) error {
+	subnetHeads, err := getSubnetChainNotify(ctx, subnetAddr, api)
 	if err != nil {
 		return err
 	}
-	if validatedBlocksNumber < 2 || validatedBlocksNumber > 100 {
-		return xerrors.New("wrong validated blocks number")
+	if n < 2 || n > 100 || m > n {
+		return fmt.Errorf("wrong blocks number")
 	}
 
 	// ChainNotify returns channel with chain head updates.
 	// First message is guaranteed to be of len == 1, and type == 'current'.
 	// Without forks we can expect that its len always to be 1.
 	initHead := <-subnetHeads
 	if len(initHead) < 1 {
-		return xerrors.New("empty chain head")
+		return fmt.Errorf("empty chain head")
 	}
 	currHeight := initHead[0].Val.Height()
 
-	head, err := api.SubnetChainHead(ctx, subnetAddr)
+	i := 0
+	minerMinedBlocks := 0
+	for i < n {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("closed channel")
+		case <-subnetHeads:
+			if m >= minerMinedBlocks {
+				return nil
+			}
+
+			head, err := getSubnetChainHead(ctx, subnetAddr, api)
+			if err != nil {
+				return err
+			}
+
+			height := head.Height()
+			if height < currHeight {
+				return fmt.Errorf("the current height is lower then the previous height")
+			}
+			if height == currHeight {
+				continue
+			}
+			currHeight = height
+
+			for _, b := range head.Blocks() {
+				if b.Miner == miner {
+					minerMinedBlocks++
+				}
+			}
+		}
+
+		i++
+	}
+
+	if m >= minerMinedBlocks {
+		return nil
+	}
+
+	return fmt.Errorf("failed to mine %d blocks", m)
+
+}
+
+// SubnetHeightCheckForBlocks checks `n` blocks with correct heights in the subnet will be mined.
+func SubnetHeightCheckForBlocks(ctx context.Context, n int, subnetAddr addr.SubnetID, api napi.FullNode) error {
+	heads, err := getSubnetChainNotify(ctx, subnetAddr, api)
 	if err != nil {
 		return err
 	}
-	height := head.Height()
-	if height != currHeight {
-		return xerrors.New("wrong initial block height")
+
+	select {
```

Comment:
Why is this first select block needed?

Reply:
Without that first select, currHead may be equal to newHead. That is because we could read them both at the same height.
If we call SubnetCHainHead twice we can get the same head. But if we call it after reading from the unbuffered channel we have a guarantee that the blocks will be different.